### DISCUSSION TITLE
How to Troubleshoot useEffect Not Triggering API Call on State Change in Next.js 14?

### DISCUSSION THREAD

**Discussion by ekiyak** [vercel/next.js#63996] [Help]
### Summary

Hello, I'm using Next.js 14 and facing an issue where a useEffect hook, triggered by a change in a state managed via context, is supposed to make a request to an API. Everything works as expected on the initial load, and the API request is successfully made. However, when the state (in this case, selectedDeliveryType) changes and the useEffect is triggered again, I am unable to make a request to the API. Despite trying with both fetch and axios, I can't seem to successfully make the request on the second trigger. Interestingly, the API and the fetch/axios call work fine on the initial load. The issue only arises after the state changes and useEffect is triggered again. How can I resolve this issue?

### Additional information

```js
// preOrderContext.jsx
import React, { createContext, useContext, useState } from 'react';

const PreOrderContext = createContext();

export const PreOrderProvider = ({ children }) => {
  const [selectedDeliveryType, setSelectedDeliveryType] = useState(1);

  const setDeliveryTypeHandler = (type) => {
    setSelectedDeliveryType(type);
  };

  return (
    <PreOrderContext.Provider value={{ selectedDeliveryType, setDeliveryTypeHandler }}>
      {children}
    </PreOrderContext.Provider>
  );
};

export const usePreOrder = () => useContext(PreOrderContext);

// SliderComponent.jsx
import React, { useEffect, useState } from 'react';
import { usePreOrder } from './preOrderContext';
import axios from 'axios';

const SliderComponent = () => {
  const { selectedDeliveryType } = usePreOrder();
  const [slides, setSlides] = useState([]);

  useEffect(() => {
    const fetchSlides = async () => {
      try {
        const response = await axios.get(`API_ENDPOINT?deliveryType=${selectedDeliveryType}`);
        setSlides(response.data);
      } catch (error) {
        console.error('Failed to fetch slides:', error);
      }
    };

    fetchSlides();
  }, [selectedDeliveryType]);

  return (
    <div>
      {slides.map((slide, index) => (
        <div key={index}>{slide.title}</div>
      ))}
    </div>
  );
};

export default SliderComponent;
```

**Comment by icyJoseph** [comment]
Hi,

Are you setting it to the same state? Like, `setDeliveryTypeHandler(1)`, given that it is already 1?

You could also:

```js
  useEffect(() => {
      console.count("selectedDeliveryType");
  }, [selectedDeliveryType]);
```

That'll show a counter in the browser, for every time the selectedDeliveryType variable is set to a new thing.

**Reply by ekiyak** [reply]
Thank you for your response and the helpful suggestion regarding useEffect and console.count. To clarify, we are indeed changing the selectedDeliveryType state and have confirmed this through UI indicators that reflect the state's current value. The issue arises when we attempt to trigger an API call through useEffect following a state change. The useEffect hook is successfully triggered (as we've verified through additional console logs and effects), but the subsequent API call (attempted with both fetch and axios) fails to execute after the initial load, even though the state change is confirmed and selectedDeliveryType is being updated with a new value. This problem persists across attempts to directly or indirectly trigger the API call following the state update. We're looking for insights or potential overlooked aspects that might be causing this issue, considering the state change and useEffect trigger are functioning as expected.

**Reply by icyJoseph** [reply]
Maybe you need to cancel the ongoing request when the dependency changes. Use an abort controller. Axios supports the abort controller signal if I remember.

Then you'd return a clean up function to abort the signal. I wrote an article about that here: https://medium.com/@icjoseph/using-react-to-understand-abort-controllers-eb10654485df

Though that doesn't sound like the issue... Which is why I didn't recommend it.

Can you at least see the request in the network tab?
